# Foundations of Computer Science - progetto 2021/22

- Fabrizio Cominetti [882737]
- Alessandro Fasani [837301]

You have to work on the files:
*  [Books](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/Books.csv.gz)
*  [Book ratings](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/Book-Ratings.csv.gz)
*  [Users](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/Users.csv.gz)
*  [Goodbooks books](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/goodbooks.csv.gz)
*  [Goodbooks ratings](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/goodbooks-ratings.csv.gz)

### Notes

1.    It is mandatory to use GitHub for developing the project.
1.    The project must be a jupyter notebook.
1.    There is no restriction on the libraries that can be used, nor on the Python version.
1.    To read those files, you need to use the `encoding = 'latin-1'` option.
1.    All questions on the project **must** be asked in a public channel on [Zulip](https://focs.zulipchat.com), otherwise no  answer will be given.

Per prima cosa importo i moduli necessari e carico i file all'interno del jupyter notebook.

In [2]:
import pandas as pd
import re
import gzip

### Importazione da GitHub

In [3]:
books = pd.read_csv("https://github.com/gdv/foundationsCS/raw/master/progetti/2021/Books.csv.gz", sep=';', compression='gzip', encoding='latin-1', low_memory=False)

In [4]:
books.shape

(271359, 8)

In [5]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [6]:
books_ratings = pd.read_csv("https://github.com/gdv/foundationsCS/raw/master/progetti/2021/Book-Ratings.csv.gz", sep=';', compression='gzip', encoding='latin-1')

In [7]:
books_ratings.shape

(1149780, 3)

In [8]:
books_ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [9]:
users = pd.read_csv("https://github.com/gdv/foundationsCS/raw/master/progetti/2021/Users.csv.gz", sep=';', compression='gzip', encoding='latin-1')

In [10]:
users.shape

(278858, 3)

In [11]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [12]:
# senza encoding ='latin-1' i nomi degli autori sono normali sena caratteri strani

gb_books = pd.read_csv("https://github.com/gdv/foundationsCS/raw/master/progetti/2021/goodbooks.csv.gz", compression='gzip')

In [13]:
gb_books.shape

(99, 23)

In [14]:
gb_books.head()

book_id  goodreads_book_id  best_book_id  work_id  books_count       isbn  \
0        1            2767052       2767052  2792775          272  439023483   
1        2                  3             3  4640799          491  439554934   
2        3              41865         41865  3212258          226  316015849   
3        4               2657          2657  3275794          487   61120081   
4        5               4671          4671   245494         1356  743273567   

         isbn13                      authors  original_publication_year  \
0  9.780439e+12              Suzanne Collins                     2008.0   
1  9.780440e+12  J.K. Rowling, Mary GrandPré                     1997.0   
2  9.780316e+12              Stephenie Meyer                     2005.0   
3  9.780061e+12                   Harper Lee                     1960.0   
4  9.780743e+12          F. Scott Fitzgerald                     1925.0   

                             original_title  ... ratings_count  \
0                          The Hunger Games  ...       4780653   
1  Harry Potter and the Philosopher's Stone  ...       4602479   
2                                  Twilight  ...       3866839   
3                     To Kill a Mockingbird  ...       3198671   
4                          The Great Gatsby  ...       2683664   

  work_ratings_count  work_text_reviews_count  ratings_1  ratings_2  \
0            4942365                   155254      66715     127936   
1            4800065                    75867      75504     101676   
2            3916824                    95009     456191     436802   
3            3340896                    72586      60427     117415   
4            2773745                    51992      86236     197621   

   ratings_3  ratings_4  ratings_5  \
0     560092    1481305    2706317   
1     455024    1156318    3011543   
2     793319     875073    1355439   
3     446835    1001952    1714267   
4     606158     936012     947718   

                                           image_url  \
0  https://images.gr-assets.com/books/1447303603m...   
1  https://images.gr-assets.com/books/1474154022m...   
2  https://images.gr-assets.com/books/1361039443m...   
3  https://images.gr-assets.com/books/1361975680m...   
4  https://images.gr-assets.com/books/1490528560m...   

                                     small_image_url  
0  https://images.gr-assets.com/books/1447303603s...  
1  https://images.gr-assets.com/books/1474154022s...  
2  https://images.gr-assets.com/books/1361039443s...  
3  https://images.gr-assets.com/books/1361975680s...  
4  https://images.gr-assets.com/books/1490528560s...  

[5 rows x 23 columns]

In [15]:
gb_books.columns

Index(['book_id', 'goodreads_book_id', 'best_book_id', 'work_id',
       'books_count', 'isbn', 'isbn13', 'authors', 'original_publication_year',
       'original_title', 'title', 'language_code', 'average_rating',
       'ratings_count', 'work_ratings_count', 'work_text_reviews_count',
       'ratings_1', 'ratings_2', 'ratings_3', 'ratings_4', 'ratings_5',
       'image_url', 'small_image_url'],
      dtype='object')

In [16]:
gb_ratings = pd.read_csv("https://github.com/gdv/foundationsCS/raw/master/progetti/2021/goodbooks-ratings.csv.gz", compression='gzip', encoding='latin-1')

In [17]:
gb_ratings.shape

(99, 3)

In [18]:
gb_ratings.head()

user_id  book_id  rating
0        1      258       5
1        2     4081       4
2        2      260       5
3        2     9296       5
4        2     2318       3

### 1. Normalize the location field of *Users* dataset, splitting into city, region, country.

_soluzione 1_

Per 'splittare' la stringa 'Location' in tre differenti colonne del nostro dataframe possiamo sfruttare il fatto che i campi che ci interessa suddividere sono separati dalla virgola ',', e quindi utilizzare il metodo split. Il parametro 'expand', che ci permette di costruire tre differenti colonne mentre, in caso non ne facessimo uso, il comando costruirebbe una singola lista.

In [19]:
users['Location'].dtype

dtype('O')

In [20]:
users[['city', 'region', 'country']] = users['Location'].str.split(',', 2, expand=True)

In [21]:
users.head()

User-ID                            Location   Age         city  \
0        1                  nyc, new york, usa   NaN          nyc   
1        2           stockton, california, usa  18.0     stockton   
2        3     moscow, yukon territory, russia   NaN       moscow   
3        4           porto, v.n.gaia, portugal  17.0        porto   
4        5  farnborough, hants, united kingdom   NaN  farnborough   

             region          country  
0          new york              usa  
1        california              usa  
2   yukon territory           russia  
3          v.n.gaia         portugal  
4             hants   united kingdom

_soluzione 2_

Users["City"]=""
Users["Region"]=""
Users["Country"]=""

Users["City"]= Users["Location"].str.extract("(\w+[\.-_!]*\s*\w*[\.-_!]*\s*\w*),s*")

Users["Region"]= Users["Location"].str.extract(",\s*(\w+[\.-_!]*\s*\w*[\.-_!]*\s*\w*),\s*")

Users["Country"]= Users["Location"].str.extract(",\s*(\w+[\.-_!]*\s*\w*[\.-_!]*\s*\w*)$")

Controllo l'eventuale presenza di valori nulla dopo aver effettuato la split

In [22]:
users['city'].isnull().sum()

0

In [23]:
users['region'].isnull().sum()

1

In [24]:
users[users['region'].isnull()]

User-ID       Location   Age           city region country
134376   134377  lawrenceville  30.0  lawrenceville   None    None

In [25]:
users['country'].isnull().sum()

2

In [26]:
users[users['country'].isnull()]

User-ID             Location   Age           city     region country
29419     29420  st. louis, missouri  68.0      st. louis   missouri    None
134376   134377        lawrenceville  30.0  lawrenceville       None    None

Ulteriore check per verificare righe con molteplici elementi o righe con valori location errati.

In [27]:
users[users['city'].str.match("\w+\,\w+")]

Empty DataFrame
Columns: [User-ID, Location, Age, city, region, country]
Index: []

In [28]:
users[users['region'].str.match("\w+\,\w+", na=False)]

Empty DataFrame
Columns: [User-ID, Location, Age, city, region, country]
Index: []

In [29]:
users[users['country'].str.match("\w+\,\w+", na=False)]

User-ID                                           Location   Age  \
58849     58850                       accra, ghana,gh,accra, ghana   NaN   
68644     68645             asmara, eritrea,maekel,asmera, eritrea   NaN   
108496   108497  90 lawerances, ave apt 2,dorchester,ma 02121, ...  29.0   
207314   207315  nieuwegein, utrecht,netherlands,europe, nether...   NaN   
235181   235182  bangalore, #341,4main,6cross,2phase,r`nagar,m`...   NaN   
243263   243264  hno:25-21/6,eastanandbagh,malkajgiri,hyderabad...   NaN   
262288   262289  hno:25-21/6,eastanadbagh,malakajgiri,hyderabad...  25.0   
274065   274066  51 tramo st., marulas,kawit,cavite, n/a, phili...   NaN   

                 city         region  \
58849           accra          ghana   
68644          asmara        eritrea   
108496  90 lawerances      ave apt 2   
207314     nieuwegein        utrecht   
235181      bangalore           #341   
243263    hno:25-21/6  eastanandbagh   
262288    hno:25-21/6   eastanadbagh   
274065   51 tramo st.        marulas   

                                                  country  
58849                                     gh,accra, ghana  
68644                              maekel,asmera, eritrea  
108496            dorchester,ma 02121, massachusetts, usa  
207314                    netherlands,europe, netherlands  
235181  4main,6cross,2phase,r`nagar,m`nagar,b`lore-10,...  
243263                    malkajgiri,hyderabad, ap, india  
262288                   malakajgiri,hyderabad, ap, india  
274065                     kawit,cavite, n/a, philippines

In [30]:
users_errors = users
users_errors['numero_campi_inseriti'] = users_errors['Location'].str.count(",")
users_errors[users_errors['numero_campi_inseriti'] > 2]

User-ID                                           Location   Age  \
435         436            ivanhoe, melbourne, victoria, australia  53.0   
814         815       davis bay, sechelt, british columbia, canada  43.0   
840         841                   saint charles, mo, missouri, usa  24.0   
956         957            santiago, santiago,metropolitana, chile   NaN   
1459       1460                mansoura, dakahlia, mansoura, egypt  34.0   
...         ...                                                ...   ...   
277717   277718          colorado springs, co 80901, colorado, usa  62.0   
277753   277754                           north bend,, oregon, usa  60.0   
278028   278029  mansfield, nottinghamshire, england, united ki...  31.0   
278214   278215                   nelson, motueka, nz, new zealand  28.0   
278808   278809          norfolk, england, england, united kingdom  49.0   

                    city            region                    country  \
435              ivanhoe         melbourne        victoria, australia   
814            davis bay           sechelt   british columbia, canada   
840        saint charles                mo              missouri, usa   
956             santiago          santiago       metropolitana, chile   
1459            mansoura          dakahlia            mansoura, egypt   
...                  ...               ...                        ...   
277717  colorado springs          co 80901              colorado, usa   
277753        north bend                                  oregon, usa   
278028         mansfield   nottinghamshire    england, united kingdom   
278214            nelson           motueka            nz, new zealand   
278808           norfolk           england    england, united kingdom   

        numero_campi_inseriti  
435                         3  
814                         3  
840                         3  
956                         3  
1459                        3  
...                       ...  
277717                      3  
277753                      3  
278028                      3  
278214                      3  
278808                      3  

[1508 rows x 7 columns]

### 2. For each book in the *Books* dataset, compute its average rating.

Sfruttiamo la presenza del campo ISBN in entrambi i dataset per effettuare una merge tra books e books_ratings. Fatto ciò eseguiamo una group by per calcolare il valore medio richiesto.

Eseguiamo una outer join, precisamente left outer join.

In [31]:
books_all = pd.merge(books, books_ratings, left_on='ISBN', right_on='ISBN', how='left')
books_all.head()

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   
2  0002005018         Clara Callan  Richard Bruce Wright                2001   
3  0002005018         Clara Callan  Richard Bruce Wright                2001   
4  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
2    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
3    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   
4    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0002005018.0...   
3  http://images.amazon.com/images/P/0002005018.0...   
4  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  User-ID  Book-Rating  
0  http://images.amazon.com/images/P/0195153448.0...      2.0          0.0  
1  http://images.amazon.com/images/P/0002005018.0...      8.0          5.0  
2  http://images.amazon.com/images/P/0002005018.0...  11400.0          0.0  
3  http://images.amazon.com/images/P/0002005018.0...  11676.0          8.0  
4  http://images.amazon.com/images/P/0002005018.0...  41385.0          0.0

In [32]:
books_all.shape

(1032344, 10)

In [33]:
books_AVG_rating = books_all.groupby(["ISBN", "Book-Title"])[["Book-Rating"]].mean()
books_AVG_rating.head()

Book-Rating
ISBN       Book-Title                                                     
0000913154 The Way Things Work: An Illustrated Encyclopedi...          8.0
0001010565 Mog's Christmas                                             0.0
0001046438 Liar                                                        9.0
0001046713 Twopence to Cross the Mersey                                0.0
000104687X T.S. Eliot Reading \The Wasteland\" and Other P...          6.0

In [34]:
books_AVG_rating.shape

(271359, 1)

### 3. For each book in the *GoodBooks* dataset, compute its average rating.

Poniamo in evidenza l'attributo 'average_rating' presente all'interno del dataset gb_books.

In [35]:
gb_books.columns

Index(['book_id', 'goodreads_book_id', 'best_book_id', 'work_id',
       'books_count', 'isbn', 'isbn13', 'authors', 'original_publication_year',
       'original_title', 'title', 'language_code', 'average_rating',
       'ratings_count', 'work_ratings_count', 'work_text_reviews_count',
       'ratings_1', 'ratings_2', 'ratings_3', 'ratings_4', 'ratings_5',
       'image_url', 'small_image_url'],
      dtype='object')

In [36]:
gb_books[['book_id', 'isbn', 'original_title', 'title', 'average_rating']]

book_id        isbn                            original_title  \
0         1   439023483                          The Hunger Games   
1         2   439554934  Harry Potter and the Philosopher's Stone   
2         3   316015849                                  Twilight   
3         4    61120081                     To Kill a Mockingbird   
4         5   743273567                          The Great Gatsby   
..      ...         ...                                       ...   
94       95   375751513                The Picture of Dorian Gray   
95       96   345803507                        Fifty Shades Freed   
96       97   393970124                                   Dracula   
97       98   307269981               Flickan som lekte med elden   
98       99  1612130585                       Fifty Shades Darker   

                                                title  average_rating  
0             The Hunger Games (The Hunger Games, #1)            4.34  
1   Harry Potter and the Sorcerer's Stone (Harry P...            4.44  
2                             Twilight (Twilight, #1)            3.57  
3                               To Kill a Mockingbird            4.25  
4                                    The Great Gatsby            3.89  
..                                                ...             ...  
94                         The Picture of Dorian Gray            4.06  
95              Fifty Shades Freed (Fifty Shades, #3)            3.88  
96                                            Dracula            3.98  
97     The Girl Who Played with Fire (Millennium, #2)            4.22  
98             Fifty Shades Darker (Fifty Shades, #2)            3.87  

[99 rows x 5 columns]

### 4. Merge together all rows sharing the same book title, author and publisher. We will call the resulting datset `merged books`. The books that have not been merged together will not appear in `merged books`.

In [37]:
books.columns

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')

In [38]:
books[books['Publisher'].isnull()]

ISBN       Book-Title      Book-Author Year-Of-Publication  \
128890  193169656X      Tyrant Moon  Elaine Corvidae                2002   
129037  1931696993  Finders Keepers  Linnea Sinclair                2001   

       Publisher                                        Image-URL-S  \
128890       NaN  http://images.amazon.com/images/P/193169656X.0...   
129037       NaN  http://images.amazon.com/images/P/1931696993.0...   

                                              Image-URL-M  \
128890  http://images.amazon.com/images/P/193169656X.0...   
129037  http://images.amazon.com/images/P/1931696993.0...   

                                              Image-URL-L  
128890  http://images.amazon.com/images/P/193169656X.0...  
129037  http://images.amazon.com/images/P/1931696993.0...

In [39]:
grouped_books = books.groupby(['Book-Title', 'Book-Author', 'Publisher'], as_index=False)

In [40]:
len(grouped_books)

266130

In [41]:
grouped_books.head()

ISBN                                         Book-Title  \
0       0195153448                                Classical Mythology   
1       0002005018                                       Clara Callan   
2       0060973129                               Decision in Normandy   
3       0374157065  Flu: The Story of the Great Influenza Pandemic...   
4       0393045218                             The Mummies of Urumchi   
...            ...                                                ...   
271354  0440400988                         There's a Bat in Bunk Five   
271355  0525447644                            From One to One Hundred   
271356  006008667X  Lily Dale : The True Story of the Town that Ta...   
271357  0192126040                        Republic (World's Classics)   
271358  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                 Book-Author Year-Of-Publication  \
0         Mark P. O. Morford                2002   
1       Richard Bruce Wright                2001   
2               Carlo D'Este                1991   
3           Gina Bari Kolata                1999   
4            E. J. W. Barber                1999   
...                      ...                 ...   
271354        Paula Danziger                1988   
271355            Teri Sloat                1991   
271356      Christine Wicker                2004   
271357                 Plato                1996   
271358   Christopher  Biffle                2000   

                                               Publisher  \
0                                Oxford University Press   
1                                  HarperFlamingo Canada   
2                                        HarperPerennial   
3                                   Farrar Straus Giroux   
4                             W. W. Norton &amp; Company   
...                                                  ...   
271354                   Random House Childrens Pub (Mm)   
271355                                      Dutton Books   
271356                                HarperSanFrancisco   
271357                           Oxford University Press   
271358  McGraw-Hill Humanities/Social Sciences/Languages   

                                              Image-URL-S  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271354  http://images.amazon.com/images/P/0440400988.0...   
271355  http://images.amazon.com/images/P/0525447644.0...   
271356  http://images.amazon.com/images/P/006008667X.0...   
271357  http://images.amazon.com/images/P/0192126040.0...   
271358  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-M  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271354  http://images.amazon.com/images/P/0440400988.0...   
271355  http://images.amazon.com/images/P/0525447644.0...   
271356  http://images.amazon.com/images/P/006008667X.0...   
271357  http://images.amazon.com/images/P/0192126040.0...   
271358  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-L  
0       http://images.amazon.com/images/P/0195153448.0...  
1       http://images.amazon.com/images/P/0002005018.0...  
2       http://images.amazon.com/images/P/0060973129.0...  
3       http://images.amazon.com/images/P/0374157065.0...  
4       http://images.ama

A questo punto condizioniamo sul fatto che debbano esistere due o più record all'interno dei vari raggruppamenti.

In [42]:
at_least_two = grouped_books.filter(lambda x: len(x)>1).sort_values(by='Book-Title')
at_least_two.head()

ISBN                                         Book-Title  \
156341  1565920317                          !%@ (A Nutshell handbook)   
75637   1565920465                          !%@ (A Nutshell handbook)   
140618  0792276833  'A Hell of a Place to Lose a Cow': An American...   
158204  0792277295  'A Hell of a Place to Lose a Cow': An American...   
179167  1566196264  10,000 dreams interpreted: A dictionary of dreams   

                    Book-Author Year-Of-Publication  \
156341            Donnalyn Frey                1993   
75637             Donnalyn Frey                1994   
140618              Tim Brookes                2000   
158204              Tim Brookes                2001   
179167  Gustavus Hindman Miller                1995   

                        Publisher  \
156341                   O'Reilly   
75637                    O'Reilly   
140618        National Geographic   
158204        National Geographic   
179167  Barnes &amp; Nobles Books   

                                              Image-URL-S  \
156341  http://images.amazon.com/images/P/1565920317.0...   
75637   http://images.amazon.com/images/P/1565920465.0...   
140618  http://images.amazon.com/images/P/0792276833.0...   
158204  http://images.amazon.com/images/P/0792277295.0...   
179167  http://images.amazon.com/images/P/1566196264.0...   

                                              Image-URL-M  \
156341  http://images.amazon.com/images/P/1565920317.0...   
75637   http://images.amazon.com/images/P/1565920465.0...   
140618  http://images.amazon.com/images/P/0792276833.0...   
158204  http://images.amazon.com/images/P/0792277295.0...   
179167  http://images.amazon.com/images/P/1566196264.0...   

                                              Image-URL-L  
156341  http://images.amazon.com/images/P/1565920317.0...  
75637   http://images.amazon.com/images/P/1565920465.0...  
140618  http://images.amazon.com/images/P/0792276833.0...  
158204  http://images.amazon.com/images/P/0792277295.0...  
179167  http://images.amazon.com/images/P/1566196264.0...

In [43]:
len(at_least_two)

9954

Ri-eseguiamo la groupby.nunique() per trovare i singoli libri. (Ovviamente i risultanti sono quei libri che hanno diverse edizioni e che avevano righe diverse precedentemente).

In [44]:
merged_books = at_least_two.groupby(['Book-Title', 'Book-Author', 'Publisher'], as_index=False)
merged_books.nunique()[['Book-Title', 'Book-Author', 'Publisher']]

Book-Title  \
0                             !%@ (A Nutshell handbook)   
1     'A Hell of a Place to Lose a Cow': An American...   
2     10,000 dreams interpreted: A dictionary of dreams   
3                                      101 Famous Poems   
4                    15 Houseplants Even You Can't Kill   
...                                                 ...   
4720                                                Zia   
4721                                                Zia   
4722                                Zimmermann Telegram   
4723  Zoids Chaotic Century (Zoids: Chaotic Century ...   
4724         ��berleben in der W�¼ste Danakil.   

                  Book-Author                       Publisher  
0               Donnalyn Frey                        O'Reilly  
1                 Tim Brookes             National Geographic  
2     Gustavus Hindman Miller       Barnes &amp; Nobles Books  
3                 Roy J. Cook  McGraw-Hill/Contemporary Books  
4                   Joe Elder               Berkley Pub Group  
...                       ...                             ...  
4720             Scott O'Dell               Laurel-Leaf Books  
4721             Scott O'Dell                  Yearling Books  
4722          Barbara Tuchman                Ballantine Books  
4723           Michiro Ueyama                      Viz Comics  
4724    R�¼diger Nehberg                           Piper  

[4725 rows x 3 columns]

### 5. For each book in `merged books` compute its average rating.

The average is computed considering all books in `books` that have been merged.

Prendo in considerazione i due dataset creati in precedenza, ovvero at_least_two e books_all.

In [45]:
merged_books2 = pd.merge(at_least_two, books_all, on="ISBN")
merged_books2.head()

ISBN                                       Book-Title_x  \
0  1565920317                          !%@ (A Nutshell handbook)   
1  1565920465                          !%@ (A Nutshell handbook)   
2  0792276833  'A Hell of a Place to Lose a Cow': An American...   
3  0792276833  'A Hell of a Place to Lose a Cow': An American...   
4  0792276833  'A Hell of a Place to Lose a Cow': An American...   

   Book-Author_x Year-Of-Publication_x          Publisher_x  \
0  Donnalyn Frey                  1993             O'Reilly   
1  Donnalyn Frey                  1994             O'Reilly   
2    Tim Brookes                  2000  National Geographic   
3    Tim Brookes                  2000  National Geographic   
4    Tim Brookes                  2000  National Geographic   

                                       Image-URL-S_x  \
0  http://images.amazon.com/images/P/1565920317.0...   
1  http://images.amazon.com/images/P/1565920465.0...   
2  http://images.amazon.com/images/P/0792276833.0...   
3  http://images.amazon.com/images/P/0792276833.0...   
4  http://images.amazon.com/images/P/0792276833.0...   

                                       Image-URL-M_x  \
0  http://images.amazon.com/images/P/1565920317.0...   
1  http://images.amazon.com/images/P/1565920465.0...   
2  http://images.amazon.com/images/P/0792276833.0...   
3  http://images.amazon.com/images/P/0792276833.0...   
4  http://images.amazon.com/images/P/0792276833.0...   

                                       Image-URL-L_x  \
0  http://images.amazon.com/images/P/1565920317.0...   
1  http://images.amazon.com/images/P/1565920465.0...   
2  http://images.amazon.com/images/P/0792276833.0...   
3  http://images.amazon.com/images/P/0792276833.0...   
4  http://images.amazon.com/images/P/0792276833.0...   

                                        Book-Title_y  Book-Author_y  \
0                          !%@ (A Nutshell handbook)  Donnalyn Frey   
1                          !%@ (A Nutshell handbook)  Donnalyn Frey   
2  'A Hell of a Place to Lose a Cow': An American...    Tim Brookes   
3  'A Hell of a Place to Lose a Cow': An American...    Tim Brookes   
4  'A Hell of a Place to Lose a Cow': An American...    Tim Brookes   

  Year-Of-Publication_y          Publisher_y  \
0                  1993             O'Reilly   
1                  1994             O'Reilly   
2                  2000  National Geographic   
3                  2000  National Geographic   
4                  2000  National Geographic   

                                       Image-URL-S_y  \
0  http://images.amazon.com/images/P/1565920317.0...   
1  http://images.amazon.com/images/P/1565920465.0...   
2  http://images.amazon.com/images/P/0792276833.0...   
3  http://images.amazon.com/images/P/0792276833.0...   
4  http://images.amazon.com/images/P/0792276833.0...   

                                       Image-URL-M_y  \
0  http://images.amazon.com/images/P/1565920317.0...   
1  http://images.amazon.com/images/P/1565920465.0...   
2  http://images.amazon.com/images/P/0792276833.0...   
3  http://images.amazon.com/images/P/0792276833.0...   
4  http://images.amazon.com/images/P/0792276833.0...   

                                       Image-URL-L_y   User-ID  Book-Rating  
0  http://images.amazon.com/images/P/1565920317.0...   87141.0          6.0  
1  http://images.amazon.com/images/P/1565920465.0...  271245.0          0.0  
2  http://images.amazon.com/images/P/0792276833.0...   79186.0          0.0  
3  http://images.amazon.com/images/P/0792276833.0...  114216.0          0.0  
4  http://images.amazon.com/images/P/0792276833.0...  201058.0          0.0

In [46]:
len(merged_books2)

92899

In [47]:
merged_books_avg_rating = merged_books2.groupby(['ISBN', 'Book-Title_x', 'Book-Author_x', 'Publisher_x'], as_index=False)[['Book-Title_x', 'Book-Author_x', 'Publisher_x', 'Book-Rating']].mean().sort_values(by='Book-Title_x')
merged_books_avg_rating.head()

ISBN                                       Book-Title_x  \
8749  1565920465                          !%@ (A Nutshell handbook)   
8748  1565920317                          !%@ (A Nutshell handbook)   
7481  0792276833  'A Hell of a Place to Lose a Cow': An American...   
7486  0792277295  'A Hell of a Place to Lose a Cow': An American...   
8756  1566196264  10,000 dreams interpreted: A dictionary of dreams   

                Book-Author_x                Publisher_x  Book-Rating  
8749            Donnalyn Frey                   O'Reilly     0.000000  
8748            Donnalyn Frey                   O'Reilly     6.000000  
7481              Tim Brookes        National Geographic     0.000000  
7486              Tim Brookes        National Geographic     3.400000  
8756  Gustavus Hindman Miller  Barnes &amp; Nobles Books     6.666667

In [48]:
merged_books_avg_rating[merged_books_avg_rating['Book-Title_x'] == "101 Famous Poems"]

ISBN      Book-Title_x Book-Author_x  \
7615  0809250969  101 Famous Poems   Roy J. Cook   
7617  0809288346  101 Famous Poems   Roy J. Cook   
7616  0809288311  101 Famous Poems   Roy J. Cook   

                         Publisher_x  Book-Rating  
7615  McGraw-Hill/Contemporary Books     0.000000  
7617  McGraw-Hill/Contemporary Books     4.333333  
7616  McGraw-Hill/Contemporary Books     5.000000

In [49]:
len(merged_books_avg_rating)

9954

### 6. For each book in `merged books` compute the minimum and maximum of the average ratings over all corresponding books in the `books` dataset.

Hence for each book in `merged books` we will have exactly two values (a minimum and a maximum)

_soluzione 1_

In [50]:
min_max = merged_books_avg_rating
min_max.groupby(['Book-Title_x', 'Book-Author_x', 'Publisher_x'])['Book-Rating'].aggregate(['min', 'max', 'count'])

min  \
Book-Title_x                                       Book-Author_x           Publisher_x                                 
!%@ (A Nutshell handbook)                          Donnalyn Frey           O'Reilly                         0.000000   
'A Hell of a Place to Lose a Cow': An American ... Tim Brookes             National Geographic              0.000000   
10,000 dreams interpreted: A dictionary of dreams  Gustavus Hindman Miller Barnes &amp; Nobles Books        6.666667   
101 Famous Poems                                   Roy J. Cook             McGraw-Hill/Contemporary Books   0.000000   
15 Houseplants Even You Can't Kill                 Joe Elder               Berkley Pub Group                0.000000   
...                                                                                                              ...   
Zia                                                Scott O'Dell            Laurel-Leaf Books                1.666667   
                                                                           Yearling Books                   0.000000   
Zimmermann Telegram                                Barbara Tuchman         Ballantine Books                 0.000000   
Zoids Chaotic Century (Zoids: Chaotic Century (... Michiro Ueyama          Viz Comics                      10.000000   
��berleben in der W�¼ste Danakil.         R�¼diger Nehberg   Piper                            2.666667   

                                                                                                                 max  \
Book-Title_x                                       Book-Author_x           Publisher_x                                 
!%@ (A Nutshell handbook)                          Donnalyn Frey           O'Reilly                         6.000000   
'A Hell of a Place to Lose a Cow': An American ... Tim Brookes             National Geographic              3.400000   
10,000 dreams interpreted: A dictionary of dreams  Gustavus Hindman Miller Barnes &amp; Nobles Books        7.250000   
101 Famous Poems                                   Roy J. Cook             McGraw-Hill/Contemporary Books   5.000000   
15 Houseplants Even You Can't Kill                 Joe Elder               Berkley Pub Group                0.000000   
...                                                                                                              ...   
Zia                                                Scott O'Dell            Laurel-Leaf Books                3.428571   
                                                                           Yearling Books                   2.000000   
Zimmermann Telegram                                Barbara Tuchman         Ballantine Books                 4.000000   
Zoids Chaotic Century (Zoids: Chaotic Century (... Michiro Ueyama          Viz Comics                      10.000000   
��berleben in der W�¼ste Danakil.         R�¼diger Nehberg   Piper                            5.000000   

                                                                                                           count  
Book-Title_x                                       Book-Author_x           Publisher_x                            
!%@ (A Nutshell handbook)                          Donnalyn Frey           O'Reilly                            2  
'A Hell of a Place to Lose a Cow': An American ... Tim Brookes             National Geographic                 2  
10,000 dreams interpreted: A dictionary of dreams  Gustavus Hindman Miller Barnes &amp; Nobles Books           2  
101 Famous Poems                                   Roy J. Cook             McGraw-Hill/Contemporary Books      3  
15 Houseplants Even You Can't Kill                 Joe Elder               Berkley Pub Group                   2  
...                                                                                                          ...  
Zia                                                Scott O'Dell        

_soluzione 2_

merged_books_avgrt_reset = merged_books_avg_rating.reset_index()
merged_books_avgrt_reset.head()

merged_books_avgrt_reset['Max'] = merged_books_avgrt_reset.groupby(['Book-Title_x', 'Book-Author_x', 'Publisher_x'])['Book-Rating'].transform(max)

merged_books_avgrt_reset['Min'] = merged_books_avgrt_reset.groupby(['Book-Title_x', 'Book-Author_x', 'Publisher_x'])['Book-Rating'].transform(min)

merged_books_avgrt_reset['Count'] = merged_books_avgrt_reset.groupby(['Book-Title_x', 'Book-Author_x', 'Publisher_x'])['Book-Rating'].transform(len)

merged_books_avgrt_reset.head()

### 7. For each book in `goodbooks`, compute the list of its authors. Assuming that the number of reviews with a text (column `work_text_reviews_count`) is split equally among all authors, find for each authors the total number of reviews with a text. We will call this quantity the *shared number of reviews with a text*.

Gli autori del medesimo libro sono separati da una virgola, utilizzeremo dunque la virgola come conteggio del numero di autori presenti per ogni libro.

In [51]:
gb_books['number_authors'] = gb_books['authors'].str.count(',')
gb_books.head(15)

book_id  goodreads_book_id  best_book_id   work_id  books_count  \
0         1            2767052       2767052   2792775          272   
1         2                  3             3   4640799          491   
2         3              41865         41865   3212258          226   
3         4               2657          2657   3275794          487   
4         5               4671          4671    245494         1356   
5         6           11870085      11870085  16827462          226   
6         7               5907          5907   1540236          969   
7         8               5107          5107   3036731          360   
8         9                960           960   3338963          311   
9        10               1885          1885   3060926         3455   
10       11              77203         77203   3295919          283   
11       12           13335037      13335037  13155899          210   
12       13               5470          5470    153313          995   
13       14               7613          7613   2207778          896   
14       15              48855         48855   3532896          710   

          isbn        isbn13  \
0    439023483  9.780439e+12   
1    439554934  9.780440e+12   
2    316015849  9.780316e+12   
3     61120081  9.780061e+12   
4    743273567  9.780743e+12   
5    525478817  9.780525e+12   
6    618260307  9.780618e+12   
7    316769177  9.780317e+12   
8   1416524797  9.781417e+12   
9    679783261  9.780680e+12   
10  1594480001  9.781594e+12   
11    62024035  9.780062e+12   
12   451524934  9.780452e+12   
13   452284244  9.780452e+12   
14   553296981  9.780553e+12   

                                              authors  \
0                                     Suzanne Collins   
1                         J.K. Rowling, Mary GrandPré   
2                                     Stephenie Meyer   
3                                          Harper Lee   
4                                 F. Scott Fitzgerald   
5                                          John Green   
6                                      J.R.R. Tolkien   
7                                       J.D. Salinger   
8                                           Dan Brown   
9                                         Jane Austen   
10                                    Khaled Hosseini   
11                                      Veronica Roth   
12            George Orwell, Erich Fromm, Celâl Üster   
13                                      George Orwell   
14  Anne Frank, Eleanor Roosevelt, B.M. Mooyaart-D...   

    original_publication_year  \
0                      2008.0   
1                      1997.0   
2                      2005.0   
3                      1960.0   
4                      1925.0   
5                      2012.0   
6                      1937.0   
7                      1951.0   
8                      2000.0   
9                      1813.0   
10                     2003.0   
11                     2011.0   
12                     1949.0   
13                     1945.0   
14                     1947.0   

                                       original_title  ... work_ratings_count  \
0                                    The Hunger Games  ...            4942365   
1            Harry Potter and the Philosopher's Stone  ...            4800065   
2                                            Twilight  ...            3916824   
3                               To Kill a Mockingbird  ...            3340896   
4                                    The Great Gatsby  ...            2773745   
5                              The Fault in Our Stars  ...            2478609   
6                  The Hobbit or There and Back Again  ...            2196809   
7                              The Catcher in the Rye  ...            2120637   
8                                    Angels & Demons   ...            2078754   
9                                 Pride and Prejudice  ...            2191465   
10                   

In [52]:
gb_books['number_authors'] = gb_books['number_authors'] + 1
gb_books.head(15)

book_id  goodreads_book_id  best_book_id   work_id  books_count  \
0         1            2767052       2767052   2792775          272   
1         2                  3             3   4640799          491   
2         3              41865         41865   3212258          226   
3         4               2657          2657   3275794          487   
4         5               4671          4671    245494         1356   
5         6           11870085      11870085  16827462          226   
6         7               5907          5907   1540236          969   
7         8               5107          5107   3036731          360   
8         9                960           960   3338963          311   
9        10               1885          1885   3060926         3455   
10       11              77203         77203   3295919          283   
11       12           13335037      13335037  13155899          210   
12       13               5470          5470    153313          995   
13       14               7613          7613   2207778          896   
14       15              48855         48855   3532896          710   

          isbn        isbn13  \
0    439023483  9.780439e+12   
1    439554934  9.780440e+12   
2    316015849  9.780316e+12   
3     61120081  9.780061e+12   
4    743273567  9.780743e+12   
5    525478817  9.780525e+12   
6    618260307  9.780618e+12   
7    316769177  9.780317e+12   
8   1416524797  9.781417e+12   
9    679783261  9.780680e+12   
10  1594480001  9.781594e+12   
11    62024035  9.780062e+12   
12   451524934  9.780452e+12   
13   452284244  9.780452e+12   
14   553296981  9.780553e+12   

                                              authors  \
0                                     Suzanne Collins   
1                         J.K. Rowling, Mary GrandPré   
2                                     Stephenie Meyer   
3                                          Harper Lee   
4                                 F. Scott Fitzgerald   
5                                          John Green   
6                                      J.R.R. Tolkien   
7                                       J.D. Salinger   
8                                           Dan Brown   
9                                         Jane Austen   
10                                    Khaled Hosseini   
11                                      Veronica Roth   
12            George Orwell, Erich Fromm, Celâl Üster   
13                                      George Orwell   
14  Anne Frank, Eleanor Roosevelt, B.M. Mooyaart-D...   

    original_publication_year  \
0                      2008.0   
1                      1997.0   
2                      2005.0   
3                      1960.0   
4                      1925.0   
5                      2012.0   
6                      1937.0   
7                      1951.0   
8                      2000.0   
9                      1813.0   
10                     2003.0   
11                     2011.0   
12                     1949.0   
13                     1945.0   
14                     1947.0   

                                       original_title  ... work_ratings_count  \
0                                    The Hunger Games  ...            4942365   
1            Harry Potter and the Philosopher's Stone  ...            4800065   
2                                            Twilight  ...            3916824   
3                               To Kill a Mockingbird  ...            3340896   
4                                    The Great Gatsby  ...            2773745   
5                              The Fault in Our Stars  ...            2478609   
6                  The Hobbit or There and Back Again  ...            2196809   
7                              The Catcher in the Rye  ...            2120637   
8                                    Angels & Demons   ...            2078754   
9                                 Pride and Prejudice  ...            2191465   
10                   

Eseguiamo una split degli autori sulle virgole:

In [53]:
gb_books['author_name'] = gb_books['authors'].str.split(',')
gb_books.head(15)

book_id  goodreads_book_id  best_book_id   work_id  books_count  \
0         1            2767052       2767052   2792775          272   
1         2                  3             3   4640799          491   
2         3              41865         41865   3212258          226   
3         4               2657          2657   3275794          487   
4         5               4671          4671    245494         1356   
5         6           11870085      11870085  16827462          226   
6         7               5907          5907   1540236          969   
7         8               5107          5107   3036731          360   
8         9                960           960   3338963          311   
9        10               1885          1885   3060926         3455   
10       11              77203         77203   3295919          283   
11       12           13335037      13335037  13155899          210   
12       13               5470          5470    153313          995   
13       14               7613          7613   2207778          896   
14       15              48855         48855   3532896          710   

          isbn        isbn13  \
0    439023483  9.780439e+12   
1    439554934  9.780440e+12   
2    316015849  9.780316e+12   
3     61120081  9.780061e+12   
4    743273567  9.780743e+12   
5    525478817  9.780525e+12   
6    618260307  9.780618e+12   
7    316769177  9.780317e+12   
8   1416524797  9.781417e+12   
9    679783261  9.780680e+12   
10  1594480001  9.781594e+12   
11    62024035  9.780062e+12   
12   451524934  9.780452e+12   
13   452284244  9.780452e+12   
14   553296981  9.780553e+12   

                                              authors  \
0                                     Suzanne Collins   
1                         J.K. Rowling, Mary GrandPré   
2                                     Stephenie Meyer   
3                                          Harper Lee   
4                                 F. Scott Fitzgerald   
5                                          John Green   
6                                      J.R.R. Tolkien   
7                                       J.D. Salinger   
8                                           Dan Brown   
9                                         Jane Austen   
10                                    Khaled Hosseini   
11                                      Veronica Roth   
12            George Orwell, Erich Fromm, Celâl Üster   
13                                      George Orwell   
14  Anne Frank, Eleanor Roosevelt, B.M. Mooyaart-D...   

    original_publication_year  \
0                      2008.0   
1                      1997.0   
2                      2005.0   
3                      1960.0   
4                      1925.0   
5                      2012.0   
6                      1937.0   
7                      1951.0   
8                      2000.0   
9                      1813.0   
10                     2003.0   
11                     2011.0   
12                     1949.0   
13                     1945.0   
14                     1947.0   

                                       original_title  ...  \
0                                    The Hunger Games  ...   
1            Harry Potter and the Philosopher's Stone  ...   
2                                            Twilight  ...   
3                               To Kill a Mockingbird  ...   
4                                    The Great Gatsby  ...   
5                              The Fault in Our Stars  ...   
6                  The Hobbit or There and Back Again  ...   
7                              The Catcher in the Rye  ...   
8                                    Angels & Demons   ...   
9                                 Pride and Prejudice  ...   
10                                   The Kite Runner   ...   
11                                          Divergent  ...   
12                               Nineteen Eighty-Four  ...   
13                         Animal Farm: A Fa

Utilizziamo il metodo explode per dividere le righe che presentano più di un autore.

In [54]:
gb_books_authors = gb_books.explode('author_name')
gb_books_authors.head()

book_id  goodreads_book_id  best_book_id  work_id  books_count       isbn  \
0        1            2767052       2767052  2792775          272  439023483   
1        2                  3             3  4640799          491  439554934   
1        2                  3             3  4640799          491  439554934   
2        3              41865         41865  3212258          226  316015849   
3        4               2657          2657  3275794          487   61120081   

         isbn13                      authors  original_publication_year  \
0  9.780439e+12              Suzanne Collins                     2008.0   
1  9.780440e+12  J.K. Rowling, Mary GrandPré                     1997.0   
1  9.780440e+12  J.K. Rowling, Mary GrandPré                     1997.0   
2  9.780316e+12              Stephenie Meyer                     2005.0   
3  9.780061e+12                   Harper Lee                     1960.0   

                             original_title  ... work_text_reviews_count  \
0                          The Hunger Games  ...                  155254   
1  Harry Potter and the Philosopher's Stone  ...                   75867   
1  Harry Potter and the Philosopher's Stone  ...                   75867   
2                                  Twilight  ...                   95009   
3                     To Kill a Mockingbird  ...                   72586   

  ratings_1  ratings_2  ratings_3  ratings_4  ratings_5  \
0     66715     127936     560092    1481305    2706317   
1     75504     101676     455024    1156318    3011543   
1     75504     101676     455024    1156318    3011543   
2    456191     436802     793319     875073    1355439   
3     60427     117415     446835    1001952    1714267   

                                           image_url  \
0  https://images.gr-assets.com/books/1447303603m...   
1  https://images.gr-assets.com/books/1474154022m...   
1  https://images.gr-assets.com/books/1474154022m...   
2  https://images.gr-assets.com/books/1361039443m...   
3  https://images.gr-assets.com/books/1361975680m...   

                                     small_image_url  number_authors  \
0  https://images.gr-assets.com/books/1447303603s...               1   
1  https://images.gr-assets.com/books/1474154022s...               2   
1  https://images.gr-assets.com/books/1474154022s...               2   
2  https://images.gr-assets.com/books/1361039443s...               1   
3  https://images.gr-assets.com/books/1361975680s...               1   

       author_name  
0  Suzanne Collins  
1     J.K. Rowling  
1    Mary GrandPré  
2  Stephenie Meyer  
3       Harper Lee  

[5 rows x 25 columns]

In [55]:
gb_books_authors['splitted_reviews'] = gb_books_authors['work_text_reviews_count'] / gb_books['number_authors']
gb_books_authors.head(15)

book_id  goodreads_book_id  best_book_id   work_id  books_count  \
0         1            2767052       2767052   2792775          272   
1         2                  3             3   4640799          491   
1         2                  3             3   4640799          491   
2         3              41865         41865   3212258          226   
3         4               2657          2657   3275794          487   
4         5               4671          4671    245494         1356   
5         6           11870085      11870085  16827462          226   
6         7               5907          5907   1540236          969   
7         8               5107          5107   3036731          360   
8         9                960           960   3338963          311   
9        10               1885          1885   3060926         3455   
10       11              77203         77203   3295919          283   
11       12           13335037      13335037  13155899          210   
12       13               5470          5470    153313          995   
12       13               5470          5470    153313          995   

          isbn        isbn13                                  authors  \
0    439023483  9.780439e+12                          Suzanne Collins   
1    439554934  9.780440e+12              J.K. Rowling, Mary GrandPré   
1    439554934  9.780440e+12              J.K. Rowling, Mary GrandPré   
2    316015849  9.780316e+12                          Stephenie Meyer   
3     61120081  9.780061e+12                               Harper Lee   
4    743273567  9.780743e+12                      F. Scott Fitzgerald   
5    525478817  9.780525e+12                               John Green   
6    618260307  9.780618e+12                           J.R.R. Tolkien   
7    316769177  9.780317e+12                            J.D. Salinger   
8   1416524797  9.781417e+12                                Dan Brown   
9    679783261  9.780680e+12                              Jane Austen   
10  1594480001  9.781594e+12                          Khaled Hosseini   
11    62024035  9.780062e+12                            Veronica Roth   
12   451524934  9.780452e+12  George Orwell, Erich Fromm, Celâl Üster   
12   451524934  9.780452e+12  George Orwell, Erich Fromm, Celâl Üster   

    original_publication_year                            original_title  ...  \
0                      2008.0                          The Hunger Games  ...   
1                      1997.0  Harry Potter and the Philosopher's Stone  ...   
1                      1997.0  Harry Potter and the Philosopher's Stone  ...   
2                      2005.0                                  Twilight  ...   
3                      1960.0                     To Kill a Mockingbird  ...   
4                      1925.0                          The Great Gatsby  ...   
5                      2012.0                    The Fault in Our Stars  ...   
6                      1937.0        The Hobbit or There and Back Again  ...   
7                      1951.0                    The Catcher in the Rye  ...   
8                      2000.0                          Angels & Demons   ...   
9                      1813.0                       Pride and Prejudice  ...   
10                     2003.0                          The Kite Runner   ...   
11                     2011.0                                 Divergent  ...   
12                     1949.0                      Nineteen Eighty-Four  ...   
12                     1949.0                      Nineteen Eighty-Four  ...   

   ratings_1 ratings_2  ratings_3  ratings_4  ratings_5  \
0      66715    127936     560092    1481305    2706317   
1      75504    101676     455024    1156318    3011543   
1      75504    101676     455024    1156318    3011543   
2     456191    436802     793319     875073    1355439   
3      60427    117415     446835    1001952    1714267   
4      86236    197621     606158     936012     947718   
5   

In [56]:
gb_books_authors.groupby('author_name', as_index = False)['splitted_reviews'].sum()

author_name  splitted_reviews
0              Alan R. Clarke      27890.500000
1     B.M. Mooyaart-Doubleday       6941.666667
2                Bernard Knox       1620.200000
3                 Celâl Üster      15172.666667
4               David J. Skal       5754.333333
..                        ...               ...
105             Veronica Roth     156896.000000
106           William Golding      26886.000000
107           William Goldman      15630.000000
108       William Shakespeare       7389.000000
109               Yann Martel      42962.000000

[110 rows x 2 columns]

### 8. For each year of publication, determine the author that has the largest value of the shared number of reviews with a text.

In [57]:
gb_books_authors.columns

Index(['book_id', 'goodreads_book_id', 'best_book_id', 'work_id',
       'books_count', 'isbn', 'isbn13', 'authors', 'original_publication_year',
       'original_title', 'title', 'language_code', 'average_rating',
       'ratings_count', 'work_ratings_count', 'work_text_reviews_count',
       'ratings_1', 'ratings_2', 'ratings_3', 'ratings_4', 'ratings_5',
       'image_url', 'small_image_url', 'number_authors', 'author_name',
       'splitted_reviews'],
      dtype='object')

Poniamo subito l'attenzione verso alcuni valori anomali della variabile 'original publication year'.

In [58]:
gb_books_authors.sort_values(by='original_publication_year')[['original_title', 'original_publication_year', 'author_name', 'splitted_reviews']].head()

original_title  original_publication_year       author_name  \
78       Ὀδύσσεια                     -720.0             Homer   
78       Ὀδύσσεια                     -720.0         E.V. Rieu   
78       Ὀδύσσεια                     -720.0   Frédéric Mugler   
78       Ὀδύσσεια                     -720.0      Bernard Knox   
78       Ὀδύσσεια                     -720.0     Robert Fagles   

    splitted_reviews  
78            1620.2  
78            1620.2  
78            1620.2  
78            1620.2  
78            1620.2

In questo caso osserviamo valori corrispondenti a -720 come anno di pubblicazione. Il libro in questione è l'Odissea, pubblicato inizialmente nell'anno 1488.

_soluzione 1_

In [59]:
bestofyear = gb_books_authors.groupby(['original_publication_year', 'author_name'], as_index=False)[['author_name', 'splitted_reviews']].sum()
bestofyear.tail(15)

original_publication_year       author_name  splitted_reviews
124                     2007.0   Stephenie Meyer           35216.0
125                     2008.0        John Green           42717.0
126                     2008.0   Stephenie Meyer           84328.0
127                     2008.0   Suzanne Collins          155254.0
128                     2009.0     James Dashner           48942.0
129                     2009.0  Kathryn Stockett           78204.0
130                     2009.0   Suzanne Collins           88538.0
131                     2010.0   Suzanne Collins           96274.0
132                     2011.0        E.L. James          103489.0
133                     2011.0     Veronica Roth          101023.0
134                     2012.0        E.L. James           25287.0
135                     2012.0     Gillian Flynn          121614.0
136                     2012.0        John Green          140739.0
137                     2012.0     Veronica Roth           55873.0
138                     2015.0     Paula Hawkins           93600.0

In [60]:
bestofyear.loc[bestofyear.groupby('original_publication_year')['splitted_reviews'].idxmax()]

original_publication_year                author_name  splitted_reviews
0                       -720.0               Bernard Knox       1620.200000
5                       1595.0   Robert           Jackson       7389.000000
7                       1811.0              Ros Ballaster       3842.000000
10                      1813.0                Jane Austen      49152.000000
11                      1818.0             Maurice Hindle       6664.333333
14                      1847.0              Michael Mason      15606.000000
18                      1859.0       Hablot Knight Browne       4364.333333
21                      1868.0          Louisa May Alcott      17090.000000
22                      1884.0               Guy Cardwell       4149.333333
25                      1891.0          Jeffrey Eugenides       9823.500000
27                      1897.0              David J. Skal       5754.333333
30                      1911.0    Frances Hodgson Burnett      13054.000000
31                      1925.0        F. Scott Fitzgerald      51992.000000
32                      1932.0              Aldous Huxley      20095.000000
33                      1936.0          Margaret Mitchell      16050.000000
34                      1937.0             J.R.R. Tolkien      37653.000000
36                      1945.0              George Orwell      35472.000000
37                      1946.0         Dom Marcos Barbosa       6134.250000
41                      1947.0    B.M. Mooyaart-Doubleday       6941.666667
44                      1949.0                Celâl Üster      15172.666667
47                      1950.0                 C.S. Lewis      15186.000000
48                      1951.0              J.D. Salinger      44920.000000
49                      1952.0             Garth Williams       4348.000000
52                      1953.0               Ray Bradbury      30694.000000
54                      1954.0            William Golding      26886.000000
55                      1958.0              Marion Wiesel      11000.500000
57                      1960.0                 Harper Lee      72586.000000
58                      1964.0           Shel Silverstein      14368.000000
61                      1967.0                S.E. Hinton      22662.000000
62                      1969.0          Kurt Vonnegut Jr.      19646.000000
63                      1973.0            William Goldman      15630.000000
64                      1974.0           Shel Silverstein       9234.000000
65                      1977.0               Stephen King      14936.000000
66                      1979.0              Douglas Adams      20345.000000
67                      1985.0           Orson Scott Card      38054.000000
68                      1988.0             Alan R. Clarke      27890.500000
70                      1989.0               John Grisham       4239.000000
71                      1990.0           Michael Crichton       8143.000000
72                      1993.0                 Lois Lowry      54084.000000
73                      1995.0             Philip Pullman      14915.000000
74                      1996.0         George R.R. Martin      46205.000000
78                      1997.0              Mary GrandPré      37933.500000
82                      1998.0              Mary GrandPré      17086.000000
88                      1999.0            Stephen Chbosky      47116.000000
90                      2000.0                  Dan Brown      25112.000000
93                      2001.0                Yann Martel      42962.000000
94                      2002.0               Alice Sebold      36642.000000
100                     2003.0            Khaled Hosseini      59730.000000
103                     2004.0               Jodi Picoult      30719.000000
112                     2005.0            Stephenie Meyer      95009.000000
117                     2006.0                 Sara Gruen      55732.000000
121                     2007.0            Cassandra C

Resta il problema che in alcuni anni è presente un solo libro ma sono presenti più autori che hanno scritto lo stesso libro. Selezioniamo però solo uno tra gli autori, siccome la splitted reviews era uguale per tutti.

_soluzione 2_

Visualizzare per ogni anno il massimo valore di 'shared number of reviews with a text', indipendentemente dal numero di volte che compare l'autore nell'anno considerato.

In [61]:
bestofyear2 = gb_books_authors.loc[gb_books_authors.groupby(['original_publication_year'])['splitted_reviews'].idxmax()]
bestofyear2[['original_publication_year', 'author_name', 'splitted_reviews']].tail()

original_publication_year      author_name  splitted_reviews
16                     2009.0  Suzanne Collins           88538.0
19                     2010.0  Suzanne Collins           96274.0
11                     2011.0    Veronica Roth          101023.0
5                      2012.0       John Green          140739.0
60                     2015.0    Paula Hawkins           93600.0

### 9. Assuming that there are no errors in the ISBN fields, find the books in both datasets, and compute the difference of average rating according to the ratings and the goodratings datasets

In [62]:
books_AVG_rating.reset_index()

ISBN                                         Book-Title  \
0       0000913154  The Way Things Work: An Illustrated Encycloped...   
1       0001010565                                    Mog's Christmas   
2       0001046438                                               Liar   
3       0001046713                       Twopence to Cross the Mersey   
4       000104687X  T.S. Eliot Reading \The Wasteland\" and Other ...   
...            ...                                                ...   
271354  B000234N76                                     Falling Angels   
271355  B000234NC6  It Must've Been Something I Ate: The Return of...   
271356  B00029DGGO                        Good Wife Strikes Back, The   
271357  B0002JV9PY                               The Blockade Runners   
271358  B0002K6K8O                               The Underground City   

        Book-Rating  
0               8.0  
1               0.0  
2               9.0  
3               0.0  
4               6.0  
...             ...  
271354          0.0  
271355          0.0  
271356          0.0  
271357          0.0  
271358          0.0  

[271359 rows x 3 columns]

In [63]:
gb_books.columns

Index(['book_id', 'goodreads_book_id', 'best_book_id', 'work_id',
       'books_count', 'isbn', 'isbn13', 'authors', 'original_publication_year',
       'original_title', 'title', 'language_code', 'average_rating',
       'ratings_count', 'work_ratings_count', 'work_text_reviews_count',
       'ratings_1', 'ratings_2', 'ratings_3', 'ratings_4', 'ratings_5',
       'image_url', 'small_image_url', 'number_authors', 'author_name'],
      dtype='object')

In [64]:
merged_books_gb = pd.merge(books_AVG_rating, gb_books, left_on='ISBN', right_on='isbn', how='inner')

In [65]:
merged_books_gb.shape

(5, 26)

In [66]:
merged_books_gb['Difference_rating'] = merged_books_gb['Book-Rating'] - merged_books_gb['average_rating']

In [67]:
merged_books_gb.columns

Index(['Book-Rating', 'book_id', 'goodreads_book_id', 'best_book_id',
       'work_id', 'books_count', 'isbn', 'isbn13', 'authors',
       'original_publication_year', 'original_title', 'title', 'language_code',
       'average_rating', 'ratings_count', 'work_ratings_count',
       'work_text_reviews_count', 'ratings_1', 'ratings_2', 'ratings_3',
       'ratings_4', 'ratings_5', 'image_url', 'small_image_url',
       'number_authors', 'author_name', 'Difference_rating'],
      dtype='object')

In [68]:
merged_books_gb[['isbn', 'title', 'Difference_rating']]

isbn                                              title  \
0  014028009X          Bridget Jones's Diary (Bridget Jones, #1)   
1  014038572X                                      The Outsiders   
2  043965548X  Harry Potter and the Prisoner of Azkaban (Harr...   
3  1400032717  The Curious Incident of the Dog in the Night-Time   
4  1594480001                                    The Kite Runner   

   Difference_rating  
0           0.001852  
1           0.400674  
2          -0.996667  
3           0.963187  
4          -1.860000

### 10. Split the users dataset according to the age. One dataset contains the users with unknown age, one with age 0-14, one with age 15-24, one with age 25-34, and so on.

In [69]:
users.columns

Index(['User-ID', 'Location', 'Age', 'city', 'region', 'country',
       'numero_campi_inseriti'],
      dtype='object')

In [70]:
users.head()

User-ID                            Location   Age         city  \
0        1                  nyc, new york, usa   NaN          nyc   
1        2           stockton, california, usa  18.0     stockton   
2        3     moscow, yukon territory, russia   NaN       moscow   
3        4           porto, v.n.gaia, portugal  17.0        porto   
4        5  farnborough, hants, united kingdom   NaN  farnborough   

             region          country  numero_campi_inseriti  
0          new york              usa                      2  
1        california              usa                      2  
2   yukon territory           russia                      2  
3          v.n.gaia         portugal                      2  
4             hants   united kingdom                      2

In [71]:
users['Age'].max()

244.0

In [72]:
users['Age'].min()

0.0

In [73]:
users[users['Age'] >= 120]

User-ID                                 Location    Age          city  \
1578       1579             akure, ondo/nigeria, nigeria  231.0         akure   
8457       8458                 milano, lombardia, italy  230.0        milano   
8781       8782                 calgary, alberta, canada  239.0       calgary   
11325     11326      johannesburg, gauteng, south africa  148.0  johannesburg   
12691     12692               jacksonville, florida, usa  151.0  jacksonville   
...         ...                                      ...    ...           ...   
257429   257430  reutlingen, baden-wuerttemberg, germany  200.0    reutlingen   
261553   261554                    arak, tennessee, iran  143.0          arak   
264933   264934                    madrid, madrid, spain  140.0        madrid   
275582   275583        koblenz, rheinland-pfalz, germany  189.0       koblenz   
276047   276048                      wien, wien, austria  127.0          wien   

                     region        country  numero_campi_inseriti  
1578           ondo/nigeria        nigeria                      2  
8457              lombardia          italy                      2  
8781                alberta         canada                      2  
11325               gauteng   south africa                      2  
12691               florida            usa                      2  
...                     ...            ...                    ...  
257429   baden-wuerttemberg        germany                      2  
261553            tennessee           iran                      2  
264933               madrid          spain                      2  
275582      rheinland-pfalz        germany                      2  
276047                 wien        austria                      2  

[78 rows x 7 columns]

In [74]:
cuts= [0,14,24,34,44,54,64,74,84,94,250]
labels=["0-14","15-24","25-34","35-44","45-54","55-64","65-74","75-84","85-94",">94"]

In [75]:
users['age_range'] = pd.cut(users['Age'], bins = cuts, labels = labels)
users.head()

User-ID                            Location   Age         city  \
0        1                  nyc, new york, usa   NaN          nyc   
1        2           stockton, california, usa  18.0     stockton   
2        3     moscow, yukon territory, russia   NaN       moscow   
3        4           porto, v.n.gaia, portugal  17.0        porto   
4        5  farnborough, hants, united kingdom   NaN  farnborough   

             region          country  numero_campi_inseriti age_range  
0          new york              usa                      2       NaN  
1        california              usa                      2     15-24  
2   yukon territory           russia                      2       NaN  
3          v.n.gaia         portugal                      2     15-24  
4             hants   united kingdom                      2       NaN

In [76]:
users['age_range'] = users['age_range'].cat.add_categories('unknown').fillna('unknown')

In [77]:
users.groupby('age_range')[['User-ID']].count()

User-ID
age_range         
0-14          3897
15-24        40001
25-34        50767
35-44        32690
45-54        23152
55-64        12493
65-74         3596
75-84          615
85-94           83
>94            386
unknown     111178

In [78]:
lista_categorie = users['age_range'].cat.categories

In [79]:
diz_age_range = {}
for i in lista_categorie:
    diz_age_range[i] = users[users['age_range'] == i]

In [80]:
for i in diz_age_range.keys():
    print ("Users in range "+ i +" \n")
    print (diz_age_range[i].head())
    print("\n")

Users in range 0-14 

     User-ID                         Location   Age               city  \
10        11   melbourne, victoria, australia  14.0          melbourne   
18        19                        weston, ,  14.0             weston   
40        41          santee, california, usa  14.0             santee   
119      120  villeneuve d`ascq, nord, france  13.0  villeneuve d`ascq   
418      419    grove city, pennsylvania, usa  14.0         grove city   

            region     country  numero_campi_inseriti age_range  
10        victoria   australia                      2      0-14  
18                                                  2      0-14  
40      california         usa                      2      0-14  
119           nord      france                      2      0-14  
418   pennsylvania         usa                      2      0-14  


Users in range 15-24 

    User-ID                               Location   Age       city  \
1         2              stockton, califo

### 11. Find the books that appear only in the goodbooks datasets.

In [81]:
gb_books_only = gb_books[gb_books.isbn.isin(books.ISBN) == False]

In [82]:
gb_books_only.head(10)

book_id  goodreads_book_id  best_book_id   work_id  books_count  \
0        1            2767052       2767052   2792775          272   
1        2                  3             3   4640799          491   
2        3              41865         41865   3212258          226   
3        4               2657          2657   3275794          487   
4        5               4671          4671    245494         1356   
5        6           11870085      11870085  16827462          226   
6        7               5907          5907   1540236          969   
7        8               5107          5107   3036731          360   
8        9                960           960   3338963          311   
9       10               1885          1885   3060926         3455   

         isbn        isbn13                      authors  \
0   439023483  9.780439e+12              Suzanne Collins   
1   439554934  9.780440e+12  J.K. Rowling, Mary GrandPré   
2   316015849  9.780316e+12              Stephenie Meyer   
3    61120081  9.780061e+12                   Harper Lee   
4   743273567  9.780743e+12          F. Scott Fitzgerald   
5   525478817  9.780525e+12                   John Green   
6   618260307  9.780618e+12               J.R.R. Tolkien   
7   316769177  9.780317e+12                J.D. Salinger   
8  1416524797  9.781417e+12                    Dan Brown   
9   679783261  9.780680e+12                  Jane Austen   

   original_publication_year                            original_title  ...  \
0                     2008.0                          The Hunger Games  ...   
1                     1997.0  Harry Potter and the Philosopher's Stone  ...   
2                     2005.0                                  Twilight  ...   
3                     1960.0                     To Kill a Mockingbird  ...   
4                     1925.0                          The Great Gatsby  ...   
5                     2012.0                    The Fault in Our Stars  ...   
6                     1937.0        The Hobbit or There and Back Again  ...   
7                     1951.0                    The Catcher in the Rye  ...   
8                     2000.0                          Angels & Demons   ...   
9                     1813.0                       Pride and Prejudice  ...   

  work_text_reviews_count ratings_1  ratings_2  ratings_3  ratings_4  \
0                  155254     66715     127936     560092    1481305   
1                   75867     75504     101676     455024    1156318   
2                   95009    456191     436802     793319     875073   
3                   72586     60427     117415     446835    1001952   
4                   51992     86236     197621     606158     936012   
5                  140739     47994      92723     327550     698471   
6                   37653     46023      76784     288649     665635   
7                   44920    109383     185520     455042     661516   
8                   25112     77841     145740     458429     716569   
9                   49152     54700      86485     284852     609755   

   ratings_5                                          image_url  \
0    2706317  https://images.gr-assets.com/books/1447303603m...   
1    3011543  https://images.gr-assets.com/books/1474154022m...   
2    1355439  https://images.gr-assets.com/books/1361039443m...   
3    1714267  https://images.gr-assets.com/books/1361975680m...   
4     947718  https://images.gr-assets.com/books/1490528560m...   
5    1311871  https://images.gr-assets.com/books/1360206420m...   
6    1119718  https://images.gr-assets.com/books/1372847500m...   
7     709176  https://images.gr-assets.com/books/1398034300m...   
8     680175  https://images.gr-assets.com/books/1303390735m...   
9    1155673  https://images.gr-assets.com/books/1320399351m...   

                                     small_image_url  number_authors  \
0  https://images.gr-assets.com/books/1447303603s...               1   
1  https://images.gr-asset

In [83]:
len(gb_books_only)

94

### 12. Assuming that each pair (author, title) identifies a book, for each book find the number of times it appears in the books dataset. Which books appear the most times?

In [84]:
books['author_name'] = books['Book-Author'].str.split(',')

In [85]:
books_single_authors = books.explode('author_name')
books_single_authors.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L           author_name  
0  http://images.amazon.com/images/P/0195153448.0...    Mark P. O. Morford  
1  http://images.amazon.com/images/P/0002005018.0...  Richard Bruce Wright  
2  http://images.amazon.com/images/P/0060973129.0...          Carlo D'Este  
3  http://images.amazon.com/images/P/0374157065.0...      Gina Bari Kolata  
4  http://images.amazon.com/images/P/0393045218.0...       E. J. W. Barber

In [86]:
conteggio = books_single_authors.groupby(['Book-Title', 'author_name']).count()

In [87]:
conteggio_dec = conteggio.sort_values(by="ISBN", ascending=False)
conteggio_dec.head()

ISBN  Book-Author  \
Book-Title                     author_name                            
Little Women                   Louisa May Alcott    21           21   
Wuthering Heights              Emily Bronte         20           20   
Adventures of Huckleberry Finn Mark Twain           20           20   
Pride and Prejudice            Jane Austen          18           18   
Dracula                        Bram Stoker          16           16   

                                                  Year-Of-Publication  \
Book-Title                     author_name                              
Little Women                   Louisa May Alcott                   21   
Wuthering Heights              Emily Bronte                        20   
Adventures of Huckleberry Finn Mark Twain                          20   
Pride and Prejudice            Jane Austen                         18   
Dracula                        Bram Stoker                         16   

                                                  Publisher  Image-URL-S  \
Book-Title                     author_name                                 
Little Women                   Louisa May Alcott         21           21   
Wuthering Heights              Emily Bronte              20           20   
Adventures of Huckleberry Finn Mark Twain                20           20   
Pride and Prejudice            Jane Austen               18           18   
Dracula                        Bram Stoker               16           16   

                                                  Image-URL-M  Image-URL-L  
Book-Title                     author_name                                  
Little Women                   Louisa May Alcott           21           21  
Wuthering Heights              Emily Bronte                20           20  
Adventures of Huckleberry Finn Mark Twain                  20           20  
Pride and Prejudice            Jane Austen                 18           18  
Dracula                        Bram Stoker                 16           16

In [88]:
conteggio = conteggio.reset_index()
conteggio.loc[conteggio['ISBN'].idxmax()][['Book-Title', 'author_name']]

Book-Title          Little Women
author_name    Louisa May Alcott
Name: 114501, dtype: object

### 13. Find the author with the highest average rating according to the goodbooks datasets.

In [89]:
gb_books_authors.columns

Index(['book_id', 'goodreads_book_id', 'best_book_id', 'work_id',
       'books_count', 'isbn', 'isbn13', 'authors', 'original_publication_year',
       'original_title', 'title', 'language_code', 'average_rating',
       'ratings_count', 'work_ratings_count', 'work_text_reviews_count',
       'ratings_1', 'ratings_2', 'ratings_3', 'ratings_4', 'ratings_5',
       'image_url', 'small_image_url', 'number_authors', 'author_name',
       'splitted_reviews'],
      dtype='object')

In [90]:
author_avg_rating = gb_books_authors.groupby('author_name', as_index=False)['average_rating'].mean()
author_avg_rating

author_name  average_rating
0              Alan R. Clarke           3.820
1     B.M. Mooyaart-Doubleday           4.100
2                Bernard Knox           3.730
3                 Celâl Üster           4.140
4               David J. Skal           3.980
..                        ...             ...
105             Veronica Roth           4.155
106           William Golding           3.640
107           William Goldman           4.250
108       William Shakespeare           3.730
109               Yann Martel           3.880

[110 rows x 2 columns]

In [91]:
author_avg_rating = author_avg_rating.sort_values(by='average_rating', ascending=False)
author_avg_rating.head()

author_name  average_rating
32          Rufus Beck        4.530000
18       Mary GrandPré        4.497143
64        J.K. Rowling        4.497143
58  George R.R. Martin        4.450000
74    Kathryn Stockett        4.450000

In [92]:
author_avg_rating.loc[author_avg_rating['average_rating'].idxmax()]

author_name        Rufus Beck
average_rating           4.53
Name: 32, dtype: object